src/main/java/io/confluent/connect/elasticsearch/DataConverter.java

```diff
@@ -119,120 +134,170 @@ static Schema preProcessSchema(Schema schema) {
-            processedMap.put(preProcessValue(entry.getKey(), keySchema, newSchema.keySchema()),
-                preProcessValue(entry.getValue(), valueSchema, newValueSchema));
-          }
-          return processedMap;
-        }
-        List<Struct> mapStructs = new ArrayList<>();
-        for (Map.Entry<?, ?> entry: map.entrySet()) {
-          Struct mapStruct = new Struct(newValueSchema);
-          mapStruct.put(MAP_KEY, preProcessValue(entry.getKey(), keySchema, newValueSchema.field(MAP_KEY).schema()));
-          mapStruct.put(MAP_VALUE, preProcessValue(entry.getValue(), valueSchema, newValueSchema.field(MAP_VALUE).schema()));
-          mapStructs.add(mapStruct);
-        }
-        return mapStructs;
+        return preProcessMapValue(value, schema, newSchema);
       case STRUCT:
-        Struct struct = (Struct) value;
-        Struct newStruct = new Struct(newSchema);
-        for (Field field : schema.fields()) {
-          Object converted =  preProcessValue(struct.get(field), field.schema(), newSchema.field(field.name()).schema());
-          newStruct.put(field.name(), converted);
-        }
-        return newStruct;
+        return preProcessStructValue(value, schema, newSchema);
       default:
         return value;
     }
   }
+
+  private static Object preProcessNullValue(Schema schema) {
+    if (schema.defaultValue() != null) {
+      return schema.defaultValue();
+    }
+    if (schema.isOptional()) {
+      return null;
+    }
+    throw new DataException("null value for field that is required and has no default value");
+  }
+
+  // @returns the decoded logical value or null if this isn't a known logical type
+  private static Object preProcessLogicalValue(String schemaName, Object value) {
+    switch (schemaName) {
+      case Decimal.LOGICAL_NAME:
+        return ((BigDecimal) value).doubleValue();
+      case Date.LOGICAL_NAME:
+      case Time.LOGICAL_NAME:
+      case Timestamp.LOGICAL_NAME:
+        return value;
+      default:
+        // User-defined type or unknown built-in
+        return null;
+    }
+  }
+
+  private static Object preProcessArrayValue(Object value, Schema schema, Schema newSchema) {
+    Collection collection = (Collection) value;
+    List<Object> result = new ArrayList<>();
+    for (Object element: collection) {
+      result.add(preProcessValue(element, schema.valueSchema(), newSchema.valueSchema()));
+    }
+    return result;
+  }
+
+  private static Object preProcessMapValue(Object value, Schema schema, Schema newSchema) {
+    Schema keySchema = schema.keySchema();
+    Schema valueSchema = schema.valueSchema();
+    Schema newValueSchema = newSchema.valueSchema();
+    Map<?, ?> map = (Map<?, ?>) value;
+    if (keySchema.type() == Schema.Type.STRING) {
+      Map<Object, Object> processedMap = new HashMap<>();
+      for (Map.Entry<?, ?> entry: map.entrySet()) {
+        processedMap.put(preProcessValue(entry.getKey(), keySchema, newSchema.keySchema()),
```

Comment:
Yeah, this just carried over from existing code that was styled differently, I've reformatted here. Style stuff like this that isn't enforced by checkstyle is just going to take time & a watchful eye to get it all converted. I'm absolutely certain there is other code that wasn't touched by this diff that also isn't formatted with this new style.